metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,660 @@
+            params=params,
+        )
+        response.raise_for_status()
+        response_dict = response.json()
+        reports: List[Report] = [
+            Report(
+                id=raw_instance["id"],
+                name=raw_instance.get("name"),
+                webUrl=raw_instance.get("webUrl"),
+                embedUrl=raw_instance.get("embedUrl"),
+                description=raw_instance.get("description"),
+                pages=self._get_pages_by_report(
+                    workspace=workspace, report_id=raw_instance["id"]
+                ),
+                users=[],  # It will be fetched using Admin Fetcher based on condition
+                tags=[],  # It will be fetched using Admin Fetcher based on condition
+                dataset=workspace.datasets.get(raw_instance.get("datasetId")),
+            )
+            for raw_instance in response_dict["value"]
+        ]
+
+        return reports
+
+    def get_report(self, workspace: Workspace, report_id: str) -> Optional[Report]:
+        reports: List[Report] = self.get_reports(
+            workspace, _filter=f"id eq '{report_id}'"
+        )
+
+        if len(reports) == 0:
+            return None
+
+        return reports[0]
+
+    def get_tiles(self, workspace: Workspace, dashboard: Dashboard) -> List[Tile]:
+
+        """
+        Get the list of tiles from PowerBi for the given workspace identifier
+
+        TODO: Pagination. As per REST API doc (https://docs.microsoft.com/en-us/rest/api/power-bi/dashboards/get
+        -tiles), there is no information available on pagination
+
+        """
+
+        def new_dataset_or_report(tile_instance: Any) -> dict:
+            """
+            Find out which is the data source for tile. It is either REPORT or DATASET
+            """
+            report_fields = {
+                "dataset": (
+                    workspace.datasets.get(tile_instance.get("datasetId"))
+                    if tile_instance.get("datasetId") is not None
+                    else None
+                ),
+                "report": (
+                    self.get_report(
+                        workspace=workspace,
+                        report_id=tile_instance.get("reportId"),
+                    )
+                    if tile_instance.get("reportId") is not None
+                    else None
+                ),
+                "createdFrom": Tile.CreatedFrom.UNKNOWN,
+            }
+
+            # Tile is either created from report or dataset or from custom visualization
+            if report_fields["report"] is not None:
+                report_fields["createdFrom"] = Tile.CreatedFrom.REPORT
+            elif report_fields["dataset"] is not None or (
+                report_fields["dataset"] is None
+                and tile_instance.get("datasetId") is not None
+            ):  # Admin API is disabled and hence dataset instance is missing
+                report_fields["createdFrom"] = Tile.CreatedFrom.DATASET
```

Comment:
This seems like an important assumption. Can you elaborate in the comment about why this means that the Admin API is disabled?

Reply:
I simplified this logic and added comment